engine/resources/mount.go

```diff
@@ -0,0 +1,687 @@
+	Freq    int               `yaml:"freq"`    // dump frequency
+	PassNo  int               `yaml:"passno"`  // verification order
+
+	mount *fstab.Mount // struct representing the mount
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *MountRes) Default() engine.Res {
+	return &MountRes{
+		Options: defaultMntOps(),
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *MountRes) Validate() error {
+	var err error
+
+	// validate state
+	if obj.State != "exists" && obj.State != "absent" {
+		return fmt.Errorf("state must be 'exists', or 'absent'")
+	}
+
+	// validate type
+	fs, err := ioutil.ReadFile(procFilesystems)
+	if err != nil {
+		return errwrap.Wrapf(err, "error reading %s", procFilesystems)
+	}
+	fsSlice := strings.Fields(string(fs))
+	for i, x := range fsSlice {
+		if x == "nodev" {
+			fsSlice = append(fsSlice[:i], fsSlice[i+1:]...)
+		}
+	}
+	if obj.State != "absent" && !util.StrInList(obj.Type, fsSlice) {
+		return fmt.Errorf("type must be a valid filesystem type (see /proc/filesystems)")
+	}
+
+	// validate mountpoint
+	if strings.Contains(obj.Name(), "//") {
+		return fmt.Errorf("double slashes are not allowed in resource name")
+	}
+	if err := unix.Access(obj.Name(), unix.R_OK); err != nil {
+		return errwrap.Wrapf(err, "error validating mount point: %s", obj.Name())
+	}
+
+	// validate device
+	device, err := evalSpec(obj.Device) // eval symlink
+	if err != nil {
+		return errwrap.Wrapf(err, "error evaluating spec: %s", obj.Device)
+	}
+	if err := unix.Access(device, unix.R_OK); err != nil {
+		return errwrap.Wrapf(err, "error validating device: %s", device)
+	}
+	return nil
+}
+
+// Init runs some startup code for this resource.
+func (obj *MountRes) Init(init *engine.Init) error {
+	obj.init = init //save for later
+
+	obj.mount = &fstab.Mount{
+		Spec:    obj.Device,
+		File:    obj.Name(),
+		VfsType: obj.Type,
+		MntOps:  obj.Options,
+		Freq:    obj.Freq,
+		PassNo:  obj.PassNo,
+	}
+	return nil
+}
+
+// Close is run by the engine to clean up after the resource is done.
+func (obj *MountRes) Close() error {
+	return nil
+}
+
+// Watch listens for signals from the mount unit associated with the resource.
+// It also watch for changes to /etc/fstab, where mounts are defined.
+func (obj *MountRes) Watch() error {
+	// make sure systemd is running
+	if !systemdUtil.IsRunningSystemd() {
+		return fmt.Errorf("systemd is not running")
+	}
+
+	// establish a godbus connection
+	conn, err := util.SystemBusPrivateUsable()
+	if err != nil {
+		return errwrap.Wrapf(err, "error establishing dbus connection")
+	}
+	defer conn.Close()
+
+	// add a dbus rule to watch signals from the mount unit.
+	args := fmt.Sprintf("type='signal', path='%s', arg0='%s'",
+		dbusUnitPath+sdbus.PathBusEscape(unit.UnitNamePathEscape((obj.Name()+".mount"))),
+		dbusMountInterface,
+	)
+	if call := conn.BusObject().Call(dbusAddMatch, 0, args); call.Err != nil {
```

Comment:
dbusAddMatch should be refactored into engine/util/ package, since it's used in a few places in the resources/ namespace already.